xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-plugins/src/main/resources/xwiki-focusedplaceholder/plugin.js

```diff
@@ -0,0 +1,268 @@
+         * @param {Object} document - the DOM Document
+         */
+        removeAllPlaceholders: function(document) {
+            document.querySelectorAll("[" + ATTRIBUTE_NAME + "]").forEach(function(match) {
+                match.removeAttribute(ATTRIBUTE_NAME);
+            });
+        },
+
+        /**
+         * Checks if an element should be considered empty
+         *
+         * @param {Object} element - the DOM Element
+         * @return {bool}
+         */
+        isEmpty: function(element) {
+
+            // Consider the length of a text node as its number of children
+            if (element.nodeName === "#text") {
+                return !Boolean(element.length);
+            }
+            var i;
+            for (i = 0; i < element.childNodes.length; i++) {
+                var child = element.childNodes[i];
+                // Child is not empty if it doesn't have a nodeName
+                if (child.nodeName === undefined) {
+                    return false;
+                }
+
+                // Child is not considered empty if it is not ignored if empty
+                if (!this.editor.config.focusedplaceholder.ignoreIfEmpty.includes(child.nodeName.toLowerCase())) {
+                    return false;
+                }
+
+                // Child is not considered empty if it is not empty
+                if (!this.isEmpty(child)) {
+                    return false;
+                }
+            }
+
+            return true;
+        },
+
+        /**
+         * Finds the first configured ancestor of a DOM element
+         *
+         * @param {Object} element
+         * @return {Object} - The first configured ancestor
+         */
+        getFirstConfiguredAncestor: function(element) {
+
+            // No parents
+            if (element === null) {
+                return element;
+            }
+            if (element.parentNode === null) {
+                return element.parentNode;
+            }
+
+            // Configured element
+            if (element.tagName !== undefined &&
+                this.editor.config.focusedplaceholder.placeholder[element.tagName.toLowerCase()] !== undefined) {
+                return element;
+            }
+
+            return this.getFirstConfiguredAncestor(element.parentNode);
+
+        },
+
+        /**
+         * Updates the placeholder when the caret moves
+         *
+         * @param {CKEDITOR.dom.range}
+         */
+        updatePlaceholder: function(range) {
+            // Clear previous placeholders
+            this.removeAllPlaceholders(range.document.$);
+
+            // No placeholder when a selection is occuring
+            if (!range.collapsed) {
+                return null;
+            }
+
+            // Find the matching ancestor
+            var ancestor = this.getFirstConfiguredAncestor(range.getCommonAncestor().$);
+            if (ancestor === null) {
+                ancestor = range.getCommonAncestor().$;
+            }
```

Comment:
This needs a comment. It's not clear to me why you need the fallback.

Reply:
Currently, the body placeholder is configured by default so an ancestor will be found.
But if the configuration changes, this is not guaranteed.
I replaced the replacement value by an empty return because if no configured ancestor is found no placeholder will be inserted anyways.
I also added a comment as you requested.